### PR TITLE
 unix.RemoveDirectories: optimize and isolate cases

### PR DESCRIPTION
**Pull request by larham**
Previously, RemoveDirectories and RemoveFiles used the unix command
"rm -rf", but this is inefficient for huge numbers of files.
Also, these functions accepted any globbed path.

Instead, use "rsync" to optimize deletion of files in a directory.
On a DCA using 1 million files, this increased speed by about 3x.

Also, this commit breaks up the different use-cases of deletion into
separate methods, adding methods RemoveDirectoryContents() and RemoveFile()
and RemoveGlob() to help isolate the assumptions of each case and
optimize for them.

### CODE REVIEW

File: gpMgmt/sbin/gprepairmirrorseg.py [hunk 1]
```diff
@@ -443,10 +435,10 @@ def __init__(self
     if is_gprepairmirrorseg_running():
         logger.error('gprepairmirrorseg is already running.  Only one instance')
         logger.error('of gprepairmirrorseg is allowed at a time.')
-        remove_pid = False
         sys.exit(1)
     else:
         create_pid_file()
+        remove_pid = True
```
Comment (Chibin) [inline gpMgmt/sbin/gprepairmirrorseg.py]: Is one of the bug that remove_pid was actually never set to True whatsoever?

File: gpMgmt/bin/gppylib/commands/unix.py [hunk 2]
```diff
@@ -520,40 +520,98 @@ def __init__(self, name, fromStr, toStr, file, ctxt=LOCAL, remoteHost=None):
...
+        unique_dir = "/tmp/emptyForRemove%s" % uuid.uuid4()
+        cmd_str = "if [ -d %(target_dir)s ]; then " \
+                  "mkdir -p %(unique_dir)s  &&  " \
+                  "%(cmd)s -a --delete %(unique_dir)s/ %(target_dir)s/  &&  " \
+                  "rmdir %(target_dir)s %(unique_dir)s ; fi" % dict(
+                    unique_dir=unique_dir,
+                    cmd=findCmdInPath('rsync'),
+                    target_dir=directory
+        )
+        Command.__init__(self, name, cmd_str, ctxt, remoteHost)
```
Comment (nadeemg) [inline gpMgmt/bin/gppylib/commands/unix.py]: Would either of the new formats be any easier to read? 
`'{first} {last}'.format(**data)`
or
`'{first} {last}'.format(first='Hodor', last='Hodor!')`

File: gpMgmt/bin/gppylib/commands/unix.py [hunk 3]
```diff
@@ -520,40 +520,98 @@ def __init__(self, name, fromStr, toStr, file, ctxt=LOCAL, remoteHost=None):
...
-        Command.__init__(self, name, cmdStr, ctxt, remoteHost)
+        unique_dir = "/tmp/emptyForRemove%s" % uuid.uuid4()
+        cmd_str = "if [ -d %(target_dir)s ]; then " \
+                  "mkdir -p %(unique_dir)s  &&  " \
+                  "%(cmd)s -a --delete %(unique_dir)s/ %(target_dir)s/  &&  " \
+                  "rmdir %(unique_dir)s ; fi" % dict(
+                    unique_dir=unique_dir,
+                    cmd=findCmdInPath('rsync'),
+                    target_dir=directory
+        )
```
Comment (nadeemg) [inline gpMgmt/bin/gppylib/commands/unix.py]: Same comment as line ~539

File: gpMgmt/sbin/gprepairmirrorseg.py (hunk 1 above)
Comment (larham) [reply gpMgmt/sbin/gprepairmirrorseg.py]: Yeah.  Good news: this unused utility is scheduled for deletion, putting it out of its misery :)

File: gpMgmt/bin/gppylib/commands/unix.py (hunk 2 above)
Comment (larham) [reply gpMgmt/bin/gppylib/commands/unix.py]: thanks, changed

**Comment by Chibin** [comment]
@khuddlefish Some changes related to gptransfer

**Comment by Chibin** [comment]
Holding off on pushing this.
We found some syntax error case when running some bash commands with environment variables.

**Comment by larham** [comment]
The underlying issue that was causing a syntax problem had to do with how environmental variables are transmitted in bash.

It turns out that a conditional "if" statement is not ok following a env setting:

```
foo=1 echo $foo  #ok

foo=1 if [ -d /tmp ]; then echo $foo ; fi  # syntax error
#-bash: syntax error near unexpected token `then'

foo=1 && if [ -d /tmp ]; then echo $foo ; fi  # ok
```

So the solution was to modify the underlying env setting to use "&&"

**Pull request merged**